src/mailer/engine.py

```diff
@@ -117,6 +119,24 @@ def _throttle_emails():
                       "Sleeping %s seconds", EMAIL_THROTTLE)
         time.sleep(EMAIL_THROTTLE)
 
+        
+def handle_backend_exception(connection, message, err):
+    # default error handler
+    if err is socket_error or \
+        err is smtplib.SMTPSenderRefused or \
+        err is smtplib.SMTPRecipientsRefused or \
+        err is smtplib.SMTPDataError or \
+        err is smtplib.SMTPAuthenticationError:
+
+        message.defer()
+        logging.info("message deferred due to failure: %s" % err)
+        MessageLog.objects.log(message, RESULT_FAILURE, log_message=str(err))
+        action = 'deferred'
+        # Kill the connection, in case the connection itself has an error.
+        connection = None
+
+    return connection, action
```

Comment:
Not sure if we should just return a None action.  I am of the opinion that if it fails it should stay deferred. That said, it's exactly why I think it should be an over ridable function.

Reply:
We could either have `action = 'deferred'` in there unconditionally or consider mails deferred whenever the error handler was called so that it has no control over `action`. I'd happy with either of those.

Reply:
I think the default action of the default handler should be to defer anything that fails. If you want specific action based on error then write specific rules for your use case (as the PR allows for).

Reply:
Sounds good. If we allow returning an `action` value other than `deferred`, what would that be — `sent` or something else? Can you think of an example where a user of django-mailer would want to return `sent` or something else? I'm asking because we could have other, simpler, design options here maybe so I'd like to understand what scenarios we anticipate and want to support. What do you think?

Reply:
Well personally I would like to see connection error's not being deferred. They are typically transient and when the condition clears you would typically want to send the mail that didn't get sent. A hard bounce however, while deferred, it would be handy to set that as "bounce" perhaps and setup notifications to site mods to investigate/remove offending email address. Perhaps a different status for soft vs hard bounces..etc...

Reply:
> Well personally I would like to see connection error's not being deferred. They are typically transient and when the condition clears you would typically want to send the mail that didn't get sent. A hard bounce however, while deferred, it would be handy to set that as "bounce" perhaps and setup notifications to site mods to investigate/remove offending email address. Perhaps a different status for soft vs hard bounces..etc...

I don't fully understand yet:
Are you saying that in some error cases you'd want to have `django-mailer` retry the same mail with a re-established connection *right now* while in others you'd want to stop processing of the mail for now but have `django-mailer` get back to that message *some time later* and that case is what we call "to defer"?

Reply:
Yes and no.  I think it can be easily summed up with: I think it's best to let the developer decide how to handle mail failures since it's such a mission critical component. By forcing only 2 states: sent and deferred, we are limiting what site ops are able to do to manage/mitigate failures and reporting of said failures. 

Again for my use case I will eventually want to take the response from my transactional provider and use that to help maintain a DB of clean, working email addresses so a "soft" and "hard" status would be used to manage bounces. Having it just say 'deferred' as the status does not provide any detail as to why that email ended up being deferred.

Reply:
On review this would be a lot of work because of the way priorities are setup and used out of the box. 

That said being able to override the handler would allow for signals or other hooks to be sent to notify site ops of status of certain errors over others. The default handler as we discussed would simply set the message to deferred.

Reply:
> Again for my use case I will eventually want to take the response from my transactional provider and use that to help maintain a DB of clean, working email addresses so a "soft" and "hard" status would be used to manage bounces.

You'll be storing that soft-or-hard-failure information in a Django model *outside* of django-mailer, correct?

Reply:
It would seem that's the easiest solution given how the priority field is being used. I still see value on returning sent or deferred from the error_handler as a custom handler could correct the condition and send the email or it could not. This could be inferred by a bool response but I think explicit is better and possibly more flexible for future.

Reply:
> I still see value on returning sent or deferred from the error_handler as a custom handler could correct the condition and send the email or it could not.

Yes, got that, my point is only about `bool` versus free-form string here.

> This could be inferred by a bool response but I think explicit is better and possibly more flexible for future.

I have made my point, I won't insist. People can already build crashing error handlers so if they return strings other than `"sent"` and `"deferred"` the crash at `counts[action] += 1` is on them, not us, alright.